### PR TITLE
Add 'open with' to the explorer context menu

### LINKED ISSUE #94347: Add Open With... to explorer context menu
re https://github.com/microsoft/vscode/issues/93767

* I have a custom editor for JSON document
* In the editor's secondary title menu I see the 'Reopen with...' command
* I would expect something similar in the explorer's context menu

Screenshot 2020-04-03 at 09 22 15

### PR DESCRIPTION
**Pull request by mjbvz**
Fixes #94347
For #94408

This change primarly adds a new `Open with...` entry to the explorer context menu. To do this however, I had to make a few other changes:

- Add a new explorer context key for available editors
- Moved the editor select prompt into a new function called `openEditorWith`
- Use `openEditorWith` for the new `open with` explorer command as well as for the existing `reopen with` command

### CODE REVIEW

**Review by bpasero — APPROVED** [review]

**Comment by isidorn** [comment]
@mjbvz hi Matt. Thanks for this PR. Code changes look good overall. 
* Please explain the context key more in the documenation in code. `ExplorerResourceAvailableEditorsContext` is too generic name and it is not explained anywhere. Please use a more precise and clear name (every explorer resource can be opened in a editor thus just from the name I am pretty puzled when is this true).
* When I execute this command I am provided with a quick pick to choose between the two editors to open with

Screenshot 2020-04-22 at 09 22 35

Some feedback regarding this:
* Placeholder text should not end with `...` and should use capial cases for start of words. Look at other placholder texts for inspiration. I suggest "Select Editor for "NAME""
* Text editor should be the last option since users can already open that just by pressing enter
* The first item should be preselected
* "Marddown Preview" should be the title of that one, not this long label
* Why do users care if it is built in or not? I suggest to remove this then the options will be in one line and leaner
* If there is only Editor + 1 option consider the following: Open With should just open it automatically using that one. Idealy the name of that command would also change though I know we do not allow that

**Review by isidorn — CHANGES_REQUESTED** [review]
My comments are above.

**Comment by mjbvz** [comment]
Thanks for the feedback. I've made most of the suggested changes. Here's some context on some of the other points:

The `open with` command will only show on files that can be opened by more than one editor, such as images. When you open an image, we use a custom editor by default instead of text editor. A common use case for `open with` is to switch from a custom editor to our default text editor, which is why I like keeping the text editor first. 

The select editor quick pick shows the extension that provides each editor choice. This is important because multiple extensions may provide custom editors with similar names, such as `Markdown preview`. The string `built in` is shown for editors that come from one of our built-in extensions.

File: extensions/markdown-language-features/package.json [hunk 1]
```diff
@@ -312,7 +312,7 @@
 		"customEditors": [
 			{
 				"viewType": "vscode.markdown.preview.editor",
-				"displayName": "(Experimental) VS Code Markdown Preview",
+				"displayName": "(Experimental) Markdown Preview",
```
Comment (isidorn) [inline extensions/markdown-language-features/package.json:315]: Please put (Experimental) at the end. I think that makes more sense

File: src/vs/workbench/contrib/files/common/files.ts [hunk 2]
```diff
@@ -77,6 +77,7 @@ export const ExplorerViewletVisibleContext = new RawContextKey<boolean>('explore
 export const ExplorerFolderContext = new RawContextKey<boolean>('explorerResourceIsFolder', false);
 export const ExplorerResourceReadonlyContext = new RawContextKey<boolean>('explorerResourceReadonly', false);
 export const ExplorerResourceNotReadonlyContext = ExplorerResourceReadonlyContext.toNegated();
+export const ExplorerResourceAvailableEditorIdsContext = new RawContextKey<string>('explorerResourceAvailableEditorIds', '');
```
Comment (isidorn) [inline src/vs/workbench/contrib/files/common/files.ts:80]: Please add some comments explaining what this context key does

**Review by isidorn — CHANGES_REQUESTED** [review]
Also please preselect the option in the `Open With` quick pick so user can just press enter

**Comment by isidorn** [comment]
@mjbvz thanks for addressing some of the feedback. I did a reivew and added comments.
Also there is a merge conflict in `fileActions.ts`.

**Pull request merged**